### PR TITLE
added hyperedge option and clusters (#926)

### PR DESCRIPTION
**Pull request by lindhuber**
Libavoid: Added hyperedge option and clusters #926

Clusters can now be modelled as hierarchical nodes (with certain restrictions).
Hyperedges have to be handled by the libavoid process itself, but an option is needed to "activate" them.

Signed-off-by: Daniel Lindhuber <daniel.lindhuber@jku.at>

### CODE REVIEW

File: plugins/org.eclipse.elk.alg.libavoid/src/org/eclipse/elk/alg/libavoid/Libavoid.melk [hunk 1]
```diff
@@ -1,5 +1,6 @@
 /*******************************************************************************
  * Copyright (c) 2016, 2022 Kiel University and others.
+                 2023 Primetals Technologies Austria GmbH
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.alg.libavoid/src/org/eclipse/elk/alg/libavoid/Libavoid.melk:3]: Can you add the organization to the line above?
I.e.: 2016, 2023 Kiel University, Primetals Technologies Austria GmbH and others.

File: plugins/org.eclipse.elk.alg.libavoid/src/org/eclipse/elk/alg/libavoid/LibavoidServerCommunicator.java [hunk 2]
```diff
@@ -1,5 +1,6 @@
 /*******************************************************************************
  * Copyright (c) 2013, 2022 Kiel University and others.
+ *               2023 Primetals Technologies Austria GmbH
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.alg.libavoid/src/org/eclipse/elk/alg/libavoid/LibavoidServerCommunicator.java:3]: See comment above.

File: plugins/org.eclipse.elk.alg.libavoid/src/org/eclipse/elk/alg/libavoid/LibavoidServerCommunicator.java [hunk 3]
```diff
@@ -465,8 +472,20 @@ private void transformGraph(final ElkNode root) {
         sb.append("GRAPHEND");
         sb.append("\n");
     }
-
+
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.alg.libavoid/src/org/eclipse/elk/alg/libavoid/LibavoidServerCommunicator.java:475]: Unnecessary whitespace

File: plugins/org.eclipse.elk.alg.libavoid/src/org/eclipse/elk/alg/libavoid/LibavoidServerCommunicator.java [hunk 4]
```diff
@@ -547,22 +576,50 @@ private void libavoidPort(final ElkPort port, final int portId, final int nodeId
...
 
     private void transformNode(final ElkNode node) {
+        if (isCluster(node)) {
+        	transformCluster(node);
+        } else {        	
+        	transformNode(node, node.getX(), node.getY());
+        }
+    }
+    
+    private boolean isCluster(final ElkNode node) {
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.alg.libavoid/src/org/eclipse/elk/alg/libavoid/LibavoidServerCommunicator.java:586]: Please add Javadoc.

File: plugins/org.eclipse.elk.alg.libavoid/src/org/eclipse/elk/alg/libavoid/LibavoidServerCommunicator.java [hunk 5]
```diff
@@ -547,22 +576,50 @@ private void libavoidPort(final ElkPort port, final int portId, final int nodeId
...
+		nodeIdCounter++;
+		
+		// transfer all ports
+		for (ElkPort port : node.getPorts()) {
+			libavoidPort(port, portIdCounter, nodeIdCounter, null);
+			portIdCounter++;
+		}
+	}
+    
+    private void transformCluster(final ElkNode node) {
```
Comment (soerendomroes) [inline plugins/org.eclipse.elk.alg.libavoid/src/org/eclipse/elk/alg/libavoid/LibavoidServerCommunicator.java:613]: Please add Javadoc

**Review by soerendomroes — CHANGES_REQUESTED** [review]
The changes look good, I have, however, not tested this feature. Could you add a basic libavoid graph to the [elk-models repository](https://github.com/eclipse/elk-models/tree/master) after this PR is merged?

**Comment by soerendomroes** [comment]
@lindhuber Could you also resolve the conflicts? It seem that by adding the omitNodeMicroLayout option, your two new options cannot be automatically merged.

**Comment by lindhuber** [comment]
@soerendomroes Should be done now, thank you for your patience. :D

**Review by soerendomroes — APPROVED** [review]

**Pull request merged**